src/string/formatter.cr

```diff
@@ -22,11 +22,33 @@ struct String::Formatter
   end
 
   private def consume_percent
-    next_char
-    flags = consume_flags
-    flags = consume_width(flags)
-    flags = consume_precision(flags)
-    consume_type(flags)
+    case next_char
+    when '{'
+      next_char
+      consume_substitution
+    else
+      flags = consume_flags
+      flags = consume_width(flags)
+      flags = consume_precision(flags)
+      consume_type(flags)
+    end
+  end
+
+  private def consume_substitution
+    word_chars = [] of Char
+    while true
+      case current_char
+      when '}'
+        break
+      else
+        word_chars << current_char
+      end
+      next_char
+    end
+    if current_arg.is_a?(Hash)
+      translations = current_arg as Hash(String, String)
+      @io << translations[word_chars.join]
+    end
```

Comment:
Definitely better. Is there no way around the type checking and casting? At least one that looks nicer?